linfa-clustering/src/optics/algorithm.rs

```diff
@@ -0,0 +1,247 @@
+    /// computed.
+    pub orderings: Vec<Sample<'a, F>>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Sample<'a, F: Float> {
+    /// Index of the sample in the dataset.
+    pub index: usize,
+    /// A reference to the datum.
+    pub observation: ArrayView<'a, F, Ix1>,
+    /// The reachability distance of a sample is the distance between the point and it's cluster
+    /// core or another point whichever is larger.
+    pub reachability_distance: Option<f64>,
+    /// The distance to the nth closest point where n is the minimum points to form a cluster.
+    pub core_distance: Option<f64>,
+}
+
+#[derive(Clone)]
+struct Neighbor<'a, F: Float> {
+    /// Index of the observation in the dataset
+    index: usize,
+    /// The observation
+    observation: ArrayView<'a, F, Ix1>,
+    /// Whether this point has been processed
+    processed: bool,
+    /// The core distance
+    c_distance: Option<FloatOrd<f64>>,
+    /// The reachability distance
+    r_distance: Option<FloatOrd<f64>>,
+}
+
+impl<'a, F: Float> Neighbor<'a, F> {
+    /// Create a new neighbor
+    fn new(index: usize, observation: ArrayView<'a, F, Ix1>) -> Self {
+        Self {
+            index,
+            observation,
+            processed: false,
+            c_distance: None,
+            r_distance: None,
+        }
+    }
+
+    /// Set the core distance given the minimum points in a cluster and the points neighbors
+    fn set_core_distance(&mut self, min_pts: usize, neighbors: &[Neighbor<'a, F>]) {
+        self.c_distance = neighbors
+            .iter()
+            .nth(min_pts - 1)
+            .map(|x| FloatOrd(self.observation.l2_dist(&x.observation).unwrap()))
+            .clone();
+    }
+
+    /// Convert the neighbor to a sample for the user
+    fn sample(&self) -> Sample<'a, F> {
+        Sample {
+            index: self.index,
+            observation: self.observation.clone(),
+            reachability_distance: self.r_distance.map(|x| x.0),
+            core_distance: self.c_distance.map(|x| x.0),
+        }
+    }
+}
+
+impl<'a, F: Float> Eq for Neighbor<'a, F> {}
+
+impl<'a, F: Float> PartialEq for Neighbor<'a, F> {
+    fn eq(&self, other: &Self) -> bool {
+        self.r_distance == other.r_distance
+    }
+
+    fn ne(&self, other: &Self) -> bool {
+        !self.eq(other)
+    }
+}
+
+impl<'a, F: Float> PartialOrd for Neighbor<'a, F> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.r_distance.partial_cmp(&other.r_distance)
+    }
+}
+
+impl<'a, F: Float> Ord for Neighbor<'a, F> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.r_distance.cmp(&other.r_distance)
+    }
+}
+
+impl Optics {
+    pub fn params(min_points: usize) -> OpticsHyperParamsBuilder {
+        OpticsHyperParams::new(min_points)
+    }
+}
+
+impl<'a, F: Float, D: Data<Elem = F>> Transformer<&'a ArrayBase<D, Ix2>, OpticsResult<'a, F>>
```

Comment:
So I'm struggling a bit with the `ArrayView` implementation because `OpticsAnalysis` doesn't own the data and it complains that `find_neighbors` borrows the data. I think I might need to seriously refactor what I've done to get round the generic/lifetime hole I've dug myself into :grimacing: 

```rust
error[E0515]: cannot return value referencing function parameter `observations`
   --> algorithms/linfa-clustering/src/optics/algorithm.rs:209:9
    |
143 |         let mut points = observations
    |                          ------------ `observations` is borrowed here
...
209 |         Ok(result)
    |         ^^^^^^^^^^ returns a value referencing data owned by the current function

error[E0515]: cannot return value referencing function parameter `observations`
   --> algorithms/linfa-clustering/src/optics/algorithm.rs:209:9
    |
174 |                 &observations,
    |                 ------------- `observations` is borrowed here
...
209 |         Ok(result)
    |         ^^^^^^^^^^ returns a value referencing data owned by the current function

error[E0515]: cannot return value referencing function parameter `observations`
   --> algorithms/linfa-clustering/src/optics/algorithm.rs:209:9
    |
194 |                     let neighbors = find_neighbors(&n.observation, &observations, self.tolerance());
    |                                                                    ------------- `observations` is borrowed here
...
209 |         Ok(result)
    |         ^^^^^^^^^^ returns a value referencing data owned by the current function
```

Reply:
did you solve this? Otherwise I can take a look

Reply:
Nope didn't manage to solve it, wanted to try and do it without having to change too much of the existing code. But I might have to bite the bullet and just make some significant changes :eyes: 

I did consider most of the lifetime stuff could be removed if the `OpticsAnalysis` didn't have the row views ordered with distances but stored the index. But dunno how forward compatible/nice-to-use that will be